### PR TITLE
Add qiskit passes

### PR DESCRIPTION
**Pull request by eendebakpt**
Adds qiskit passes for:

- Removal of small gates
- Removal of diagonal gates
- Decomposition of `U` and `CX` gates

### CODE REVIEW

File: src/tests/unittests/qiskit/test_passes.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        qcd = delay_pass(qc)
+        node = list(qcd)[-1][0]
+        self.assertEqual(node.name, 'delay')
+        self.assertEqual(node.params, [40])
+
+        delay_pass = DelayPass(gate_durations={'x': 80, 'h': 40}, delay_quantum=40)
+        qc = QuantumCircuit(3)
+        qc.h(0)
+        qc.x(1)
+        qcd = delay_pass(qc)
```
Comment (QFer) [inline src/tests/unittests/qiskit/test_passes.py:128]: This pass is not tested only executed. qcd is overwritten in line 136 without testing it.

File: src/tests/unittests/qiskit/test_passes.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+from qiskit.circuit import QuantumCircuit
+from qiskit.circuit.library import CZGate, HGate, RXGate, RYGate, RZGate
+
+from qtt.qiskit.passes import (DecomposeCX, DecomposeU, DelayPass,
+                               LinearTopologyParallelPass,
+                               RemoveDiagonalGatesAfterInput,
+                               RemoveSmallRotations, SequentialPass)
+
+
+class TestQiskitPasses(unittest.TestCase):
```
Comment (QFer) [inline src/tests/unittests/qiskit/test_passes.py:14]: The tests are not testing edge cases of all passes. From code review I cannot tell if the passes are error free in edge cases.

**Review by QFer — CHANGES_REQUESTED** [review]

File: src/qtt/qiskit/passes.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+    def __init__(self, epsilon: float = 0, modulo2pi=False):
+        """Remove all small rotations from a circuit
+        Args:
+            epsilon: Threshold for rotation angle to be removed
+            modulo2pi: If True, then rotations multiples of 2pi are removed as well
+        """
+        super().__init__()
+
+        self.epsilon = epsilon
+        self._empty_dag1 = qiskit.converters.circuit_to_dag(QuantumCircuit(1))
```
[the remark this answers is not in the document]
Comment (eendebakpt) [reply src/qtt/qiskit/passes.py:33]: For efficiency we do not want to re-create a dag every time

File: src/qtt/qiskit/passes.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
+        """Run the pass on `dag`.
+        Args:
+            dag: input dag.
+        Returns:
+            Output dag with small rotations removed
+        """
+        def modulo_2pi(x):
+            x = float(x)
+            return np.mod(x + np.pi, 2 * np.pi) - np.pi
```
[the remark this answers is not in the document]
Comment (eendebakpt) [reply src/qtt/qiskit/passes.py:46]: It does not matter here, we only care whether the resulting angle is close to zero. Changing the method to leave pi the same would involve another if statement

File: src/qtt/qiskit/passes.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+                    phi = float(node.op.params[0])
+                    if self.mod2pi:
+                        phi = modulo_2pi(phi)
+                    if np.abs(phi) <= self.epsilon:
+                        dag.substitute_node_with_dag(node, self._empty_dag2)
+        return dag
+
+
+class RemoveDiagonalGatesAfterInput(TransformationPass):
+    """Remove diagonal gates (including diagonal 2Q gates) at the start of a circuit.
```
[the remark this answers is not in the document]
Comment (eendebakpt) [reply src/qtt/qiskit/passes.py:72]: No, Tofolli is not diagonal in the measurement basis.

File: src/tests/unittests/qiskit/test_passes.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
+import unittest
```
Comment (QFer) [inline src/tests/unittests/qiskit/test_passes.py:1]: Please add `__init__.py` to this qiskit directory in unittests

**Review by QFer — CHANGES_REQUESTED** [review]

**Review by QFer — APPROVED** [review]
